crates/red_knot_python_semantic/src/types/narrow.rs

```diff
@@ -27,62 +28,117 @@ use std::sync::Arc;
 /// constraint is applied to that definition, so we'd just return `None`.
 pub(crate) fn narrowing_constraint<'db>(
     db: &'db dyn Db,
-    test: Expression<'db>,
+    constraint: Constraint<'db>,
     definition: Definition<'db>,
 ) -> Option<Type<'db>> {
-    all_narrowing_constraints(db, test)
-        .get(&definition.symbol(db))
-        .copied()
+    match constraint {
+        Constraint::Expression(expression) => all_narrowing_constraints(db, expression)
+            .get(&definition.symbol(db))
+            .copied(),
+        Constraint::Pattern(pattern) => all_narrowing_constraints_for_pattern(db, pattern)
+            .get(&definition.symbol(db))
+            .copied(),
+    }
+}
+
+#[salsa::tracked(return_ref)]
+fn all_narrowing_constraints_for_pattern<'db>(
+    db: &'db dyn Db,
+    pattern: PatternConstraint<'db>,
+) -> NarrowingConstraints<'db> {
+    NarrowingConstraintsBuilder::new(db, Constraint::Pattern(pattern)).finish()
 }
 
 #[salsa::tracked(return_ref)]
 fn all_narrowing_constraints<'db>(
     db: &'db dyn Db,
-    test: Expression<'db>,
+    expression: Expression<'db>,
 ) -> NarrowingConstraints<'db> {
-    NarrowingConstraintsBuilder::new(db, test).finish()
+    NarrowingConstraintsBuilder::new(db, Constraint::Expression(expression)).finish()
 }
 
 type NarrowingConstraints<'db> = FxHashMap<ScopedSymbolId, Type<'db>>;
 
 struct NarrowingConstraintsBuilder<'db> {
     db: &'db dyn Db,
-    expression: Expression<'db>,
+    constraint: Constraint<'db>,
     constraints: NarrowingConstraints<'db>,
 }
 
 impl<'db> NarrowingConstraintsBuilder<'db> {
-    fn new(db: &'db dyn Db, expression: Expression<'db>) -> Self {
+    fn new(db: &'db dyn Db, constraint: Constraint<'db>) -> Self {
         Self {
             db,
-            expression,
+            constraint,
             constraints: NarrowingConstraints::default(),
         }
     }
 
     fn finish(mut self) -> NarrowingConstraints<'db> {
-        if let ast::Expr::Compare(expr_compare) = self.expression.node_ref(self.db).node() {
-            self.add_expr_compare(expr_compare);
+        match self.constraint {
+            Constraint::Expression(expression) => self.evaluate_expression_constraint(expression),
+            Constraint::Pattern(pattern) => self.evaluate_pattern_constraint(pattern),
         }
-        // TODO other test expression kinds
 
         self.constraints.shrink_to_fit();
         self.constraints
     }
 
+    fn evaluate_expression_constraint(&mut self, expression: Expression<'db>) {
+        let inference = infer_expression_types(self.db, expression);
```

Comment:
We should only do this if we've actually found a case below that we know how to handle, otherwise it's wasted Salsa work.